lib/sitemap/generator.rb

```diff
@@ -0,0 +1,135 @@
+        file_name:    file_name,
+      )
+      file_name
+    end
+
+    def create_sitemap_index(sitemaps)
+      @uploader.upload(
+        file_content: generate_sitemap_index_xml(sitemaps),
+        file_name:    "sitemap.xml",
+      )
+    end
+
+    # Generate a sitemap which matches the format specified in https://www.sitemaps.org/protocol.html
+    def generate_sitemap_xml(documents)
+      builder = Nokogiri::XML::Builder.new(encoding: "UTF-8") do |xml|
+        xml.urlset(xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9") do
+          documents.each do |document|
+            xml.url {
+              xml.loc document.url
+              xml.lastmod document.last_updated if document.last_updated
+              xml.priority document.priority
+            }
+          end
+        end
+      end
+      builder.to_xml
+    end
+
+    def generate_sitemap_index_xml(sitemap_filenames)
+      builder = Nokogiri::XML::Builder.new(encoding: "UTF-8") do |xml|
+        xml.sitemapindex(xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9") do
+          sitemap_filenames.each do |sitemap_filename|
+            xml.sitemap {
+              xml.loc "#{base_url}/#{SUB_DIRECTORY}/#{sitemap_filename}"
+              xml.lastmod @timestamp.strftime("%FT%T%:z")
+            }
+          end
+        end
+      end
+      builder.to_xml
+    end
+
+    def batches_of_documents
+      Enumerator::Lazy.new(0...batch_total_count) do |yielder, index|
+        yielder << [homepage] + get_documents_for_batch(index)
+      end
+    end
+
+  private
+
+    def get_documents_for_batch(batch_index)
+      @search_client.search(
+        index: index_names,
+        body: batch_documents_query(batch_index),
+      )["hits"]["hits"].map do |hit|
+        SitemapPresenter.new(hit["_source"], property_boost_calculator)
+      end
+    end
+
+    EXCLUDED_FORMATS = %w[recommended-link].freeze
+    SITEMAP_LIMIT = 25_000
+    SUB_DIRECTORY = "sitemaps".freeze
+
+    StaticDocumentPresenter = Struct.new(:url, :last_updated, :priority)
+
+    def index_names
+      SearchConfig.content_index_names + [SearchConfig.govuk_index_name]
+    end
+
+    def homepage
+      StaticDocumentPresenter.new(Plek.current.website_root + "/", nil, 0.5)
+    end
+
+    def batch_total_count
+      @batch_total_count ||= document_count.fdiv(SITEMAP_LIMIT).ceil
+    end
+
+    def document_count
+      query = {
+        query: {
+          bool: {
+            must_not: { terms: { format: EXCLUDED_FORMATS } },
+          },
+        },
+      }
+      @search_client.count(body: query, index: index_names)["count"]
+    end
+
+    def batch_documents_query(batch_index)
+      {
+        from: batch_index * SITEMAP_LIMIT,
+        size: SITEMAP_LIMIT,
+        query: {
+          bool: {
+            must_not: { terms: { format: EXCLUDED_FORMATS } },
+          },
+        },
+        post_filter: Search::FormatMigrator.new(@search_config).call,
+      }
+    end
```

Comment:
I have some concerns about this. 

1. We might need to change the index config for it to work:

    > Note that `from` + `size` can not be more than the `index.max_result_window` index setting, which defaults to 10,000.

    https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-from-size

2. We have 400,000+ documents, which is a *lot* larger than 10,000, so I wonder if increasing the window size that much would cause performance issues.

3. How will this handle new documents being inserted during sitemap generation?  We tried to do something like this to import ES2 data into ES5, and ended up with many missing documents.

Can't we use the [scroll API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-scroll)?  It's designed for processing large numbers of documents and handles insertions/deletions mid-scroll.  We could set a scroll timeout of 10m or something, which should be far more than enough to process a batch of results.

Reply:
I can look into making scroll API lazier 👌